File: android/src/main/java/app/capgo/contentsquare/ContentsquarePlugin.java
package app.capgo.contentsquare;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.webkit.WebView;

import androidx.annotation.NonNull;

import com.contentsquare.android.Contentsquare;
import com.contentsquare.android.api.CsWebViewManager;
import com.contentsquare.android.api.bridge.xpf.ExternalBridgeInterface;
import com.contentsquare.android.api.bridge.xpf.ExternalBridgeType;
import com.contentsquare.android.api.bridge.xpf.SDKState;
import com.contentsquare.android.api.bridge.xpf.SDKStateChangeType;
import com.contentsquare.android.api.bridge.xpf.XpfInterface;
import com.contentsquare.android.api.model.Transaction;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

import java.lang.reflect.Method;

@CapacitorPlugin(name = "Contentsquare")
public class ContentsquarePlugin extends Plugin {

    private ContentsquareTelemetry telemetry;
    private ExternalBridgeInterface xpfBridge;
    private ContentsquareJSInjector jsInjector;

    @Override
    public void load() {
        try {
            final WebView webView = getBridge().getWebView();
            Contentsquare.unMask(webView);
            jsInjector = new ContentsquareJSInjector(webView);

            xpfBridge = new ExternalBridgeInterface() {
                @Override
                public void notifyShouldSendInitialInsertion() {
                }

                @Override
                public void updateBridgeConfig(@NonNull final String config) {
                }

                @Override
                public void notifySrMaskingHasChanged(final boolean enabled) {
                }

                @Override
                public void notifySDKStateChanges(@NonNull final SDKStateChangeType changeType, @NonNull final SDKState sdkState) {
                }

                @Override
                public void notifyCsInAppEnabled(final boolean enabled) {
                }

                @Override
                public void notifySessionReplayEnabled(final boolean enabled) {
                }

                @NonNull
                @Override
                public ExternalBridgeType getBridgeType() {
                    return null;
                }

                @Override
                public void notifySrMaskingIndicatorHasChanged(final boolean value) {
                }
            };

            webView.post(() -> {
                CsWebViewManager.INSTANCE.injectEventTrackingInterface(webView);
                webView.reload();
                Log.d("CSLIBCAP", "WebView injected, registering XPF bridge");
                registerXpfBridge();
            });

            telemetry = new ContentsquareTelemetry();
            telemetry.setXPFType();
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "Failed to initialize plugin", exception);
        }
    }

    @Override
    protected void handleOnDestroy() {
        unregisterXpfBridge();
        super.handleOnDestroy();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void sendScreenName(final PluginCall call) {
        final String name = call.getString("name");
        if (name == null) {
            call.reject("Missing 'name' parameter.");
            return;
        }

        final String js = "window._uxa = window._uxa || [];\n"
            + "window._uxa.push([\"trackPageview\", " + JSONObject.quote(name) + "]);\n";
        jsInjector.addToJSQueue("TrackPageview: " + name, js);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void optIn(final PluginCall call) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Contentsquare.optIn(getContext());
            call.resolve();
        });
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void optOut(final PluginCall call) {
        new Handler(Looper.getMainLooper()).post(() -> {
            Contentsquare.optOut(getContext());
            call.resolve();
        });
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void sendTransaction(final PluginCall call) {
        try {
            final Float price = call.getFloat("transactionValue");
            final String currency = call.getString("transactionCurrency");
            final String transactionId = call.getString("transactionId");

            if (price == null || currency == null) {
                call.reject("Missing 'transactionValue' or 'transactionCurrency' parameter.");
                return;
            }

            if (transactionId == null || transactionId.isEmpty()) {
                Contentsquare.send(Transaction.Companion.builder(price, currency).build());
            } else {
                Contentsquare.send(Transaction.Companion.builder(price, currency).id(transactionId).build());
            }
            call.resolve();
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "sendTransaction failed", exception);
            call.reject("sendTransaction failed.", exception);
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void sendDynamicVarWithIntValue(final PluginCall call) {
        try {
            final String key = call.getString("dynVarKey");
            final Integer value = call.getInt("dynVarValue");

            if (key == null || value == null) {
                call.reject("Missing 'dynVarKey' or 'dynVarValue' parameter.");
                return;
            }

            Contentsquare.send(key, value);
            call.resolve();
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "sendDynamicVarWithIntValue failed", exception);
            call.reject("sendDynamicVarWithIntValue failed.", exception);
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void sendDynamicVarWithStringValue(final PluginCall call) {
        try {
            final String key = call.getString("dynVarKey");
            final String value = call.getString("dynVarValue");

            if (key == null || value == null) {
                call.reject("Missing 'dynVarKey' or 'dynVarValue' parameter.");
                return;
            }

            Contentsquare.send(key, value);
            call.resolve();
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "sendDynamicVarWithStringValue failed", exception);
            call.reject("sendDynamicVarWithStringValue failed.", exception);
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void onReady(final PluginCall call) {
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void setCapturedElementsSelector(final PluginCall call) {
        final String elements = call.getString("elements");
        if (elements == null) {
            call.reject("Missing 'elements' parameter.");
            return;
        }

        final String js = "window._uxa = window._uxa || [];\n"
            + "window._uxa.push([\"setCapturedElementsSelector\", " + JSONObject.quote(elements) + "]);\n";
        jsInjector.addToJSQueue("setCapturedElementsSelector: " + elements, js);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void excludeURLForReplay(final PluginCall call) {
        final String urlPattern = call.getString("url");
        if (urlPattern == null) {
            call.reject("Missing 'url' parameter.");
            return;
        }

        final String js = "window._uxa = window._uxa || [];\n"
            + "window._uxa.push([\"excludeURLforReplay\", " + urlPattern + "]);\n";
        jsInjector.addToJSQueue("excludeURLForReplay: " + urlPattern, js);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void setPIISelectors(final PluginCall call) {
        final String js = "window._uxa = window._uxa || [];\n"
            + "window._uxa.push([\"setPIISelectors\", " + call.getData().toString() + "]);\n";
        jsInjector.addToJSQueue("setPIISelectors", js);
        call.resolve();
    }

    @PluginMethod(returnType = PluginMethod.RETURN_PROMISE)
    public void collect(final PluginCall call) {
        try {
            final String name = call.getString("name");
            final String value = call.getString("value", "");

            if (telemetry == null) {
                call.reject("Telemetry is not available.");
                return;
            }

            if (name == null) {
                call.reject("Missing 'name' parameter.");
                return;
            }

            telemetry.collect(name, value);
            call.resolve();
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "collect failed", exception);
            call.reject("collect failed.", exception);
        }
    }

    private void registerXpfBridge() {
        try {
            final Method registerMethod = XpfInterface.class.getDeclaredMethod("registerExternalBridge", ExternalBridgeInterface.class);
            registerMethod.setAccessible(true);
            final Object xpfInterfaceInstance = XpfInterface.class.getDeclaredField("INSTANCE").get(null);
            registerMethod.invoke(xpfInterfaceInstance, xpfBridge);
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "registerXpfBridge failed", exception);
        }
    }

    private void unregisterXpfBridge() {
        try {
            if (xpfBridge == null) {
                return;
            }
            final Method unregisterMethod = XpfInterface.class.getDeclaredMethod("unregisterExternalBridge", ExternalBridgeInterface.class);
            unregisterMethod.setAccessible(true);
            final Object xpfInterfaceInstance = XpfInterface.class.getDeclaredField("INSTANCE").get(null);
            unregisterMethod.invoke(xpfInterfaceInstance, xpfBridge);
        } catch (Exception exception) {
            Log.e("CSLIBCAP", "unregisterXpfBridge failed", exception);
        }
    }
}

File: ios/Sources/ContentsquarePlugin/ContentsquarePlugin.swift
import Capacitor
import ContentsquareModule
import Foundation

@objc(_ExternalBridgeInterface)
protocol ExternalBridgeInterface {
    @objc func takeSnapshot(parameters: [String: Any])
    @objc func enableSessionReplay(parameters: [String: Any])
    @objc func enableAPIErrors(parameters: [String: Any])
    @objc func enableCrashReporter(parameters: [String: Any])
    @objc func setTagId(_: String)
    @objc func updateFeatureFlags(_: [[String: Any]])
    @objc func notifyCSInAppEnabled(parameters: [String: Any])
    @objc func updateBridgeConfig(_: Data)
    @objc func shouldHandleWebView(_: AnyObject)
    @objc func shouldHandleExternalBridge(_: AnyObject)
    @objc func enableUnmaskAll(parameters: [String: Any])
    @objc func notifySDKStateChanges(parameters: [String: Any])
    @objc func notifySrMaskingVisualisationEnabled(parameters: [String: Any])
}

@objc(ContentsquarePlugin)
public class ContentsquarePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ContentsquarePlugin"
    public let jsName = "Contentsquare"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "optIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "optOut", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendScreenName", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendDynamicVarWithStringValue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendDynamicVarWithIntValue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCapturedElementsSelector", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "excludeURLForReplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setPIISelectors", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "collect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "onReady", returnType: CAPPluginReturnPromise)
    ]

    private var telemetry: ContentsquareTelemetry?
    private var tagInjector: ContentsquareTagInjector?

    override public func load() {
        guard let webView = bridge?.webView else {
            NSLog("ContentsquarePlugin: WebView is not available.")
            return
        }

        Contentsquare.register(webView: webView)
        registerExternalBridge()
        telemetry = ContentsquareTelemetry()
        telemetry?.setXPFType()
        tagInjector = ContentsquareTagInjector(webView: webView)
    }

    @objc func optIn(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Contentsquare.optIn()
            call.resolve()
        }
    }

    @objc func optOut(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Contentsquare.optOut()
            call.resolve()
        }
    }

    @objc func sendScreenName(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject("Missing 'name' parameter.")
            return
        }

        guard let tagInjector else {
            call.reject("Tag injector is not available.")
            return
        }

        do {
            let script = """
            window._uxa = window._uxa || [];
            window._uxa.push(["trackPageview", \(try javaScriptStringLiteral(name))]);
            """
            tagInjector.addToJSQueue(tag: "TrackPageview: \(name)", jsToInject: script)
            call.resolve()
        } catch {
            call.reject("Failed to serialize screen name.", error.localizedDescription, error)
        }
    }

    @objc func sendTransaction(_ call: CAPPluginCall) {
        let transactionValue = call.getFloat("transactionValue", 0)
        let transactionCurrency = call.getString("transactionCurrency", "")
        let transactionId = call.getString("transactionId", "")

        Contentsquare.send(transaction: CustomerTransaction(id: transactionId, value: transactionValue, currency: transactionCurrency))
        call.resolve()
    }

    @objc func sendDynamicVarWithStringValue(_ call: CAPPluginCall) {
        let dynVarKey = call.getString("dynVarKey", "")
        let dynVarValue = call.getString("dynVarValue", "")

        Contentsquare.send(dynamicVar: DynamicVar(key: dynVarKey, value: dynVarValue))
        call.resolve()
    }

    @objc func sendDynamicVarWithIntValue(_ call: CAPPluginCall) {
        let dynVarKey = call.getString("dynVarKey", "")
        let dynVarValue = UInt32(call.getInt("dynVarValue", 0))

        Contentsquare.send(dynamicVar: DynamicVar(key: dynVarKey, value: dynVarValue))
        call.resolve()
    }

    @objc func setCapturedElementsSelector(_ call: CAPPluginCall) {
        guard let elements = call.getString("elements") else {
            call.reject("Missing 'elements' parameter.")
            return
        }

        guard let tagInjector else {
            call.reject("Tag injector is not available.")
            return
        }

        do {
            let script = """
            window._uxa = window._uxa || [];
            window._uxa.push(["setCapturedElementsSelector", \(try javaScriptStringLiteral(elements))]);
            """
            tagInjector.addToJSQueue(tag: "setCapturedElementsSelector: \(elements)", jsToInject: script)
            call.resolve()
        } catch {
            call.reject("Failed to serialize elements selector.", error.localizedDescription, error)
        }
    }

    @objc func excludeURLForReplay(_ call: CAPPluginCall) {
        guard let urlPattern = call.getString("url") else {
            call.reject("Missing 'url' parameter.")
            return
        }

        guard let tagInjector else {
            call.reject("Tag injector is not available.")
            return
        }

        do {
            let serializedPattern = try javaScriptStringLiteral(urlPattern)
            let script = """
            window._uxa = window._uxa || [];
            window._uxa.push(["excludeURLforReplay", \(serializedPattern)]);
            """

            tagInjector.addToJSQueue(tag: "excludeURLForReplay: \(urlPattern)", jsToInject: script)
            call.resolve()
        } catch {
            call.reject("Failed to serialize URL pattern.", error.localizedDescription, error)
        }
    }

    @objc func setPIISelectors(_ call: CAPPluginCall) {
        guard let piiSelectors = call.getArray("PIISelectors") else {
            call.reject("Missing 'PIISelectors' parameter.")
            return
        }

        guard let piiAttributes = call.getArray("Attributes") else {
            call.reject("Missing 'Attributes' parameter.")
            return
        }

        guard let tagInjector else {
            call.reject("Tag injector is not available.")
            return
        }

        let piiObject: [String: Any] = [
            "PIISelectors": piiSelectors,
            "Attributes": piiAttributes
        ]

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: piiObject, options: [])
            guard let jsonString = String(data: jsonData, encoding: .utf8) else {
                call.reject("Failed to encode PII configuration.")
                return
            }

            let script = """
            window._uxa = window._uxa || [];
            window._uxa.push(["setPIISelectors", \(jsonString)]);
            """
            tagInjector.addToJSQueue(tag: "setPIISelectors", jsToInject: script)
            call.resolve()
        } catch {
            call.reject("Failed to serialize PII configuration.", error.localizedDescription, error)
        }
    }

    @objc func collect(_ call: CAPPluginCall) {
        guard let telemetry else {
            call.reject("Telemetry is not available.")
            return
        }

        let name = call.getString("name", "")
        let value = call.getString("value", "")
        telemetry.collect(name: name, value: value)
        call.resolve()
    }

    @objc func onReady(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func registerExternalBridge() {
        Contentsquare.perform(
            NSSelectorFromString("_registerExternalBridgeWithParameters:"),
            with: [
                "interface": self,
                "type": 4
            ] as [String: Any]
        )
    }

    private func javaScriptStringLiteral(_ value: String) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: [value], options: [])
        guard let serialized = String(data: data, encoding: .utf8),
              serialized.count >= 2
        else {
            throw NSError(domain: "ContentsquarePlugin", code: 1)
        }

        return String(serialized.dropFirst().dropLast())
    }
}

extension ContentsquarePlugin: ExternalBridgeInterface {
    func notifySDKStateChanges(parameters: [String: Any]) {
    }

    func notifySrMaskingVisualisationEnabled(parameters: [String: Any]) {
    }

    func takeSnapshot(parameters: [String: Any]) {
    }

    func enableSessionReplay(parameters: [String: Any]) {
    }

    func enableAPIErrors(parameters: [String: Any]) {
    }

    func enableCrashReporter(parameters: [String: Any]) {
    }

    func setTagId(_: String) {
    }

    func updateFeatureFlags(_: [[String: Any]]) {
    }

    func notifyCSInAppEnabled(parameters: [String: Any]) {
    }

    func updateBridgeConfig(_: Data) {
    }

    func shouldHandleWebView(_: AnyObject) {
    }

    func shouldHandleExternalBridge(_: AnyObject) {
    }

    func enableUnmaskAll(parameters: [String: Any]) {
    }
}
